src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentSplatLoweringPass.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Components
+{
+    internal class ComponentSplatLoweringPass : ComponentIntermediateNodePassBase, IRazorOptimizationPass
+    {
+        // Run after component lowering pass
+        public override int Order => 50;
+
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+        {
+            if (!IsComponentDocument(documentNode))
+            {
+                return;
+            }
+
+            var references = documentNode.FindDescendantReferences<TagHelperDirectiveAttributeIntermediateNode>();
+            var parents = new HashSet<IntermediateNode>();
+            for (var i = 0; i < references.Count; i++)
+            {
+                parents.Add(references[i].Parent);
+            }
+
+            for (var i = 0; i < references.Count; i++)
+            {
+                var reference = references[i];
+                var node = (TagHelperDirectiveAttributeIntermediateNode)reference.Node;
+
+                if (!reference.Parent.Children.Contains(node))
+                {
+                    // This node was removed as a duplicate, skip it.
```

Comment:
When will it ever get here? I don't see anything removing duplicates.

Reply:
yeah good point.